src/consensus/dkg.rs

```diff
@@ -73,69 +84,95 @@ impl Debug for DkgKey {
 pub(crate) struct DkgVoter {
-    participant: Option<Participant>,
-    observers: HashMap<DkgKey, Observer>,
+    session: Option<Session>,
 
     // Due to the asyncronous nature of the network we might sometimes receive a DKG message before
-    // we created the corresponding `Participant` session. To avoid losing those messages, we store
-    // them in this backlog and replay them once we create the session.
+    // we created the corresponding session. To avoid losing those messages, we store them in this
+    // backlog and replay them once we create the session.
     backlog: Backlog,
 }
 
 impl Default for DkgVoter {
     fn default() -> Self {
         Self {
-            participant: None,
-            observers: HashMap::new(),
+            session: None,
             backlog: Backlog::new(),
         }
     }
 }
 
 impl DkgVoter {
-    // Starts a new DKG session as a participant.
-    pub fn start_participating(
+    // Starts a new DKG session.
+    pub fn start(
         &mut self,
         our_name: XorName,
         dkg_key: DkgKey,
         elders_info: EldersInfo,
     ) -> Vec<DkgCommand> {
-        if let Some(session) = &self.participant {
+        if let Some(session) = &self.session {
             if session.dkg_key == dkg_key && session.elders_info.is_some() {
                 trace!("DKG for {} already in progress", elders_info);
                 return vec![];
             }
         }
 
+        let index = if let Some(index) = elders_info.position(&our_name) {
+            index
+        } else {
+            error!(
+                "DKG for {} failed to start: {} is not a participant",
+                elders_info, our_name
+            );
+            return vec![];
+        };
+
         // Special case: only one participant.
         if elders_info.elders.len() == 1 {
             let secret_key_set = bls::SecretKeySet::random(0, &mut rand::thread_rng());
 
-            return vec![DkgCommand::HandleParticipationResult {
-                dkg_key,
+            return vec![DkgCommand::HandleOutcome {
                 elders_info,
-                result: Ok(DkgOutcome {
+                outcome: SectionKeyShare {
                     public_key_set: secret_key_set.public_keys(),
+                    index,
                     secret_key_share: secret_key_set.secret_key_share(0),
-                }),
+                },
             }];
         }
 
+        // Keep trying to initialize until we succeed. This should always terminate in a finite
+        // number of iterations (usually small).
+        let mut next_dkg_key = dkg_key;
+        let mut next_elders_info = elders_info;
+
+        loop {
+            match self.try_initialize(our_name, next_dkg_key, next_elders_info, index) {
+                Ok(commands) => break commands,
+                Err(elders_info) => {
+                    next_dkg_key = next_dkg_key.retry();
+                    next_elders_info = elders_info;
```

Comment:
this is bit weird. the function is asking for a DkgGen supposed about the input `elders_info`,
but here the DkgGen will be silently started about the returned `error` elders_info.

Also, I think the  `next_dkg_key.retry()` shall not be used as the dkg_key, but shall be a key deduced from the error returned elders_info ?

Also, also, if we do decide the KeyGen shall be about the error returned elders_info, then we shall at least return this back to upper caller to make it be notified ?

Reply:
The elders info returned in the `Err` is the same one that was used to start the DKG. When a DKG fails and is restarted, it uses the same participants, but different DKG key (to avoid mixing the messages). Maybe this too should be commented?

Reply:
> Maybe this too should be commented?

yes. better to make it clear.